### ISSUE TITLE
Input tensor for if node experiences redundant host <-> device memory copies

### ISSUE THREAD

**Issue by admayber** [microsoft/onnxruntime#11254]
**Describe the bug**
We have traced an image classification model that uses an if node triggered by a boolean input to dynamically enable / disable certain a particular subgraph at inference time. The subgraph holds our preprocessing stages so it is at the beginning of the onnx graph.

We are trying to leverage the CUDA EP for lower latency, and in production we will have already loaded the input image into device memory. During our profiling, however, we found that the image is being copied from the device to host memory at input binding time, and then immediately copied *back* into the device when Run() is called. These extra copies are completely redundant and, for larger images, have a huge impact on our e2e pipeline latency. (See screenshots)

Based on our experiments it seems that the specific cause is the use of an if node at the beginning of the graph which consumes the image as an input. We've included a minimal example onnx file as well as python code to generate it below.

**Urgency**
This is a severe blocking issue for some of our deliverables in the 8 - 10 week range.

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04
- ONNX Runtime installed from (source or binary): gpu binary from release
- ONNX Runtime version: 1.10-gpu
- Python version: N/A
- Visual Studio version (if applicable): N/A
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version: CUDA 11.4, cuDNN 8.2.4
- GPU model and memory: V100, 16GB

**To Reproduce**
See working code below for our setup for running inference against the provided onnx model.

See the screenshot provided for the corresponding nsys trace. You will see three memory transitions happen during the following code, as labeled in the screenshot:
1. Host-to-device: this is done in a non-ORT stream when we run `cudaMemcpy` below
2. Device-to-host: this happens at IO binding time, the model seems to want the image loaded into *host* memory even though there is no cpu work that requires the image
3. Host-to-device: happens at the Run() call, the binding copied the image to host but then the model immediately requests that it be moved back to the device

Note that we did some quick experiments inserting sleep delays to confirm where in the code flow these memory transfers happen.

```cpp
// set up sessionoptions for cuda
Ort::Env env(OrtLoggingLevel::ORT_LOGGING_LEVEL_ERROR, "eval");
OrtCUDAProviderOptions cudaOptions;
cudaOptions.gpu_mem_limit = 6L * (1024L * 1024L * 1024L);
Ort::SessionOptions sessionOptions;
sessionOptions.SetIntraOpNumThreads(6);
sessionOptions.SetGraphOptimizationLevel(GraphOptimizationLevel::ORT_ENABLE_ALL);
sessionOptions.AppendExecutionProvider_CUDA(cudaOptions);

// build model from provided sample onnx
Ort::Session icModel(env, "ifelse.onnx", sessionOptions);

// build cpu / gpu allocators
Ort::AllocatorWithDefaultOptions cpuAllocator;
Ort::MemoryInfo cpuMemoryInfo =
    Ort::MemoryInfo::CreateCpu(OrtAllocatorType::OrtArenaAllocator, OrtMemType::OrtMemTypeDefault);
Ort::MemoryInfo gpuMemoryInfo = Ort::MemoryInfo("Cuda", OrtAllocatorType::OrtDeviceAllocator, 0, OrtMemTypeDefault);

// build rectangular image with random float values using OpenCV
const size_t ImgDim = 1000;
cv::Mat img(ImgDim, ImgDim, CV_32FC3);
cv::randu(img, cv::Scalar(0), cv::Scalar(255));

// set up input metadata
std::vector<int64_t> imageInputDims{1, 3, imgDim, imgDim};
static const bool FalseValue = false;
std::vector<int64_t> controlInputDims{1};

// copy image to device memory
Ort::Allocator cudaAllocator(modelSession, gpuMemoryInfo);
size_t imageBufferSize = 3 * static_cast<size_t>(imgDim) * static_cast<size_t>(imgDim) * sizeof(float);
void* inputBuffer = cudaAllocator.Alloc(imageBufferSize);
cudaMemcpy(inputBuffer, batch.ptr<float>(), imageBufferSize, cudaMemcpyHostToDevice);

std::vector<Ort::Value> inputTensors;

// create input tensor with image in gpu memory
inputTensors.emplace_back(Ort::Value::CreateTensor<float>(
    gpuMemoryInfo,
    reinterpret_cast<float*>(inputBuffer),
    3 * static_cast<size_t>(imgDim) * static_cast<size_t>(imgDim),
    imageInputDims.data(),
    imageInputDims.size()));

// add if / else control input tensor from cpu memory
inputTensors.emplace_back(
    Ort::Value::CreateTensor<bool>(cpuMemoryInfo,
                                    const_cast<bool*>(&FalseValue),
                                    1,
                                    controlInputDims.data(),
                                    controlInputDims.size()));

// bind I/O and sync inputs
Ort::IoBinding binding(modelSession);
binding.BindInput("image", inputTensors[0]);
binding.BindInput("control", inputTensors[1]);
binding.BindOutput("output", gpuMemoryInfo);
binding.SynchronizeInputs();

// run inference
modelSession.Run(Ort::RunOptions{nullptr}, binding);
```

**Screenshots**
![image](https://user-images.githubusercontent.com/44822281/163886619-1ce2dd1e-adcd-4211-b405-090c076fc29a.png)

**Additional context**
Zip file with full nsys report (screenshotted above) and sample onnx file:
[ifelse.zip](https://github.com/microsoft/onnxruntime/files/8508459/ifelse.zip)

quick python code to generate the sample onnx:
```python
import torch.nn as nn
import torch
import onnxruntime as ort
import numpy as np

class IfElseModule(nn.Module):
    def forward(self, image, control):
        if not control:
            return image

        return torch.add(image, 10)

ifelse_module = torch.jit.script(IfElseModule())
trace_image = torch.rand((1, 224, 224, 3), dtype=torch.float32)
trace_control = torch.tensor([True])
onnx_filename = "ifelse.onnx"

torch.onnx.export(ifelse_module, (trace_image, trace_control), 
    onnx_filename, input_names=["image", "control"], output_names=["output"], dynamic_axes={"image": [1, 2, 3]}, opset_version=11)

session = ort.InferenceSession(onnx_filename)
print(session.run(["output"], {"image": np.random.default_rng().standard_normal((1, 224, 224, 3), dtype=np.float32), "control": np.array([True])}))
```

**Comment by hariharans29** [comment]
I believe I know the reason for this redundant round trip. I will investigate it but I think it is because a gap in our location planning for certain kinds of inputs (specficially a graph input that is consumed not in the main graph but in the subgraph - like in your case). Due to this gap in the planning, IOBinding thinks that the input is "required" on CPU which causes the first part of the round-trip. The second part is because the data itself is required on CUDA and this causes the copy from CPU to CUDA. 


Some follow-up questions:

1) If the round-trip copy is fixed, are you sure the latency will improve to the point where your model(s) can be shipped ?

2) Are you able to build from source to validate a fix ?

**Comment by admayber** [comment]
@hariharans29 thanks for your response, yeah that makes sense as a probable cause. Per your questions:

1. Very much so. Memory operation time is currently overwhelming the inference run time because we are making a lot of repeated inference calls against the same (fairly large) image with slightly different parameters each time
2. Yes, I've done source builds recently so this should be no issue at all (might need guidance on the specific parameters to use)

**Comment by pommedeterresautee** [comment]
I am experiencing the very same issue than described above.

I have an ONNX model with a `If` node at the very beginning of its computation graph, in the then branch there is a transformer (decoder) computation graph which supports caching and in the else branch the same model without caching support.

The output of the ONNX model are ok, but the computation is 3-4X slower when I am using the model with the If node + 2 subgraphs Vs using any of the subgraph directly.

The `If` node is conditioned on some scalar bool input.
Inputs are sent to Onnx Runtime through io binding API.

Code is available there: https://github.com/ELS-RD/transformer-deploy/blob/t5/t5.ipynb

I tested the fix you are working on (https://github.com/microsoft/onnxruntime/pull/11320) but it didn't brought any speed improvement compared to ORT 1.11 :-(

```shell
git clone --recursive https://github.com/Microsoft/onnxruntime
cd onnxruntime
git fetch origin hari/location_plan_implicit_inputs
git checkout -b hari/location_plan_implicit_inputs FETCH_HEAD
CUDACXX=/usr/local/cuda-11.5/bin/nvcc ./build.sh \
    --config Release \
    --build_wheel \
    --parallel \
    --use_cuda \
    --cuda_home /usr/local/cuda-11.5 \
    --cudnn_home /usr/lib/x86_64-linux-gnu/ \
    --skip_test
```

@hariharans29 Am I supposed to see an improvement in speed with https://github.com/microsoft/onnxruntime/pull/11320, or will there be several PR related to this issue?

**Comment by hariharans29** [comment]
@[pommedeterresautee](https://github.com/pommedeterresautee) : This is the only PR I have at the moment and it is trying to solve a vecry specific issue of having an If node at the top of the graph consuming graph inputs that is cauing round-trip copies in the OP's case. Maybe there is a different issue or combinations of multiple issues surfacing from your use-case of having an If node - it is hard to tell without inspecting the graph. I am waiting on the OP to confirm that their issue was fixed by #11320 and if it is, unfortunately, that is the only fix I have for now. 

I suggest opening a different issue with more details (a minimal ONNX model to repro at the very least) and show how you are using the ORT APIs for someone on the team to make an assessment.

**Comment by pommedeterresautee** [comment]
No issue, will do it.

Just for sharing, here is an nsight screenshot were we can see memory transfers (and that's why I think it's the same issue):

Model with if/else node:

![image](https://user-images.githubusercontent.com/1029874/165143774-74d82abf-c048-45c5-84e2-b48c19471905.png)

Without if/else (very same subgraph but no if else on top of it):

![image](https://user-images.githubusercontent.com/1029874/165144051-268f3f4e-7864-4696-8227-54cbd4e1a097.png)

**Comment by pommedeterresautee** [comment]
@hariharans29 please find the 2 models used in the script above in these GDrive links:
- ONNX model (decoder with cache): https://drive.google.com/file/d/1M3i3g-2ib8BZpu64INqI98Kw5pHY6uaZ/view?usp=sharing
- ONNX model (decoder without cache): https://drive.google.com/file/d/1-JTgAcTh-2D9QGzIFOeorcsals6tzJwc/view?usp=sharing

They are the decoder part of `T5` small flavor (from hugging face hub) exported by Pytorch to ONNX without any modification/optimization.

As explained above, the two model computation graphs are "merged" in a single one through a `If` node.
Most of the ONNX code in the script above is to manage collisions between node/axis/weight names.

The goal is to support cache/no cache setup on autoregressive models without duplicating weights (which is the usual way of doing inference on T5 when using ONNX like in https://github.com/Ki6an/fastT5). FWIW no model from Hugging Face library supports JIT script export so only tracing is possible.

Merged model outputs has been checked (compared with Pytorch output) with real data and everything is working as expected.

`If` node consumes only one input (a `bool` called "enable_cache") and all the other inputs are only consumed by the subgraphs (there are 27 inputs in the merged model). One subgraph has more input than the other (past state for the K/Vcache) which is similar to the case described in the PR and by the OP.

I edited my previous message so the script has no dependency outside onnx / onnxruntime / pytorch / numpy.

Let me know if somethings else is needed to hep for reproducibility.

**Comment by admayber** [comment]
@hariharans29 for my part, it looks like the fix in https://github.com/microsoft/onnxruntime/pull/11320 addressed the issue I was having, I'm no longer seeing extraneous round-trip memory copies and inference time has dropped dramatically as expected

**Comment by hariharans29** [comment]
@admayber : Thanks for the confirmation. I merged the PR now.

@pommedeterresautee : Since the OP confirmed that their bug is fixed with the PR, I am guessing the issue your model is experiencing (albeit the similar model structure with the `If` node) is a different one. Can you please open another issue with the details you shared here and someone will dig into it.

**Comment by pommedeterresautee** [comment]
Thank you @hariharans29, it's done.
